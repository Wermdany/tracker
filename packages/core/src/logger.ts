export interface LoggerConfig {
  isLogger: boolean;
}

export interface Logger {
  warn: (...args: any[]) => void;
  info: (...args: any[]) => void;
  error: (...args: any[]) => void;
  throwError: (base: Error | string) => never;
}

export function createLogger<N extends string>(name: N, options?: Partial<LoggerConfig>): Logger {
  const { isLogger = true } = options || {};

  const genPrefix = () => {
    return name ? "[".concat(__NAME__, "/", name, "]") : "[".concat(__NAME__, "]");
  };

  return {
    warn(...args: any[]) {
      isLogger && console.warn(genPrefix(), ...args);
    },
    info(...args: any[]) {
      isLogger && console.info(genPrefix(), ...args);
    },
    error(...args: any[]) {
      isLogger && console.error(genPrefix(), ...args);
    },
    throwError(base: Error | string): never {
      if (!(base instanceof Error)) {
        base = new Error(base);
        Object.defineProperty(base, "name", {
          value: "".concat(titleToUpperCase(__NAME__), titleToUpperCase(name), "Error")
        });
      }
      throw base;
    }
  };
}

function titleToUpperCase(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}
