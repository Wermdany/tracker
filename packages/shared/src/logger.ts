export interface Logger {
  bold(msg: string): string;
  info(...args: any[]): void;
  log(...args: any[]): void;
  warn(...args: any[]): void;
  error(...args: any[]): void;
  debug(...args: any[]): void;
  msg(...args: string[]): string;
}

type LogType = "info" | "warn" | "error";

export interface LoggerOptions {
  prefix: string;
}

interface MethodsOptions {
  prefix?: string;
  type?: LogType;
}

export function createLogger(module?: string, options: Partial<LoggerOptions> = {}): Logger {
  const { prefix = "@tracker" } = options;

  const bold = (word: string) => {
    return `\x1b[1m${word}\x1b[22m`;
  };

  const pre = module ? bold(`[${prefix}/${module}] `) : bold(`[${prefix}] `);

  const methods = (options: MethodsOptions, ...args: string[]) => {
    const { type, prefix = pre } = options;
    if (!type) {
      return [prefix].concat(args).join(" ");
    }
    console[type](prefix, ...args);
  };

  const logger: Logger = {
    bold,
    log(...args) {
      methods({ type: "info" }, ...args);
    },
    info(...args) {
      methods({ type: "info" }, ...args);
    },
    warn(...args) {
      methods({ type: "warn" }, ...args);
    },
    error(...args) {
      methods({ type: "error" }, ...args);
    },
    debug(...args) {
      const debug = "\x1B[45m\x1B[97m DEBUG \x1B[39m\x1B[49m";
      methods({ type: "info", prefix: debug }, ...args);
    },
    msg(...args) {
      return methods({}, ...args)!;
    }
  };

  return logger;
}
