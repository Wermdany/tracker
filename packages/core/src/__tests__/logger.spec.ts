import { createLogger } from "../logger";

const moduleStr = "core";

const logger = createLogger(moduleStr);

describe("test logger", () => {
  it("info", () => {
    expect(logger.info("1", "2")).toHaveConsoleInfo(moduleStr, "1 2");
    expect(logger.info("1", "2")).toHaveConsoleInfo(moduleStr, "1 2");
  });

  it("warn", () => {
    expect(logger.warn("1", "2")).toHaveConsoleWarn(moduleStr, "1 2");
    expect(logger.warn("1", "2")).toHaveConsoleWarn(moduleStr, "1 2");
  });

  it("error", () => {
    expect(logger.error("1", "2")).toHaveConsoleError(moduleStr, "1 2");
    expect(logger.error("1", "2")).toHaveConsoleError(moduleStr, "1 2");
  });
  it("config => isLogger = false", () => {
    const closeLogger = createLogger("", { isLogger: false });
    // not console
    expect(closeLogger.info("", "3", "4")).toHaveConsoleInfo(moduleStr, "1 2");
    expect(closeLogger.warn("", "3", "4")).toHaveConsoleInfo(moduleStr, "1 2");
    expect(closeLogger.error("", "3", "4")).toHaveConsoleInfo(moduleStr, "1 2");
  });

  it("throw error", () => {
    expect(() => {
      logger.throwError("has a Error");
    }).toThrowErrorMatchingSnapshot();
  });
});
