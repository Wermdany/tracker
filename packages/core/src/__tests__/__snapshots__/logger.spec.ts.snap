// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test logger throw error 1`] = `"has a Error"`;
